### PR TITLE
[Moves] Fully Implement Syrup Bomb

### PR DESCRIPTION
**Pull request by frutescens**
## What are the changes the user will see?
Syrup Bomb will have its correct secondary effect of decreasing the afflicted Pokemon's speed once per turn for three turns. 

## Why am I making these changes?
Shareholder value up!

## What are the changes from a developer perspective?
New Battler Tag for Syrup Bomb

### Screenshots/Videos

https://github.com/user-attachments/assets/830c0cf3-097d-4c4c-b1e1-e0eb6f4c8c85

## How to test the changes?
Use Syrup Bomb on a Pokemon without Bulletproof. 

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/enums/battler-tag-type.ts [hunk 1]
```diff
@@ -85,4 +85,5 @@ export enum BattlerTagType {
   TORMENT = "TORMENT",
   TAUNT = "TAUNT",
   IMPRISON = "IMPRISON",
+  SYRUP_BOMB = "SYRUP_BOMB"
```
Comment (DayKev) [inline src/enums/battler-tag-type.ts:88]:
```suggestion
  SYRUP_BOMB = "SYRUP_BOMB",
```

File: src/test/moves/syrup_bomb.test.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import { allMoves } from "#app/data/move";
+import { Moves } from "#enums/moves";
+import { Species } from "#enums/species";
+import { Abilities } from "#enums/abilities";
+import { BattlerTagType } from "#enums/battler-tag-type";
+import { Stat } from "#enums/stat";
+import GameManager from "#test/utils/gameManager";
+import Phaser from "phaser";
+import { afterEach, beforeAll, beforeEach, describe, expect, test, vi } from "vitest";
```
Comment (DayKev) [inline src/test/moves/syrup_bomb.test.ts:9]:
```suggestion
import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
```

File: src/test/moves/syrup_bomb.test.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      .startingLevel(30)
+      .enemyLevel(100)
+      .moveset([Moves.SYRUP_BOMB, Moves.SPLASH])
+      .enemyMoveset(Moves.SPLASH);
+    vi.spyOn(allMoves[Moves.SYRUP_BOMB], "accuracy", "get").mockReturnValue(100);
+  });
+
+  //Bulbapedia Reference: https://bulbapedia.bulbagarden.net/wiki/syrup_bomb_(move)
+
+  test("decreases the target Pokemon's speed stat once per turn for 3 turns",
```
Comment (DayKev) [inline src/test/moves/syrup_bomb.test.ts:39]:
```suggestion
  it("decreases the target Pokemon's speed stat once per turn for 3 turns",
```

File: src/test/moves/syrup_bomb.test.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      expect(targetPokemon.getStatStage(Stat.SPD)).toBe(-2);
+
+      game.move.select(Moves.SPLASH);
+      await game.toNextTurn();
+      expect(targetPokemon.getTag(BattlerTagType.SYRUP_BOMB)).toBeUndefined();
+      expect(targetPokemon.getStatStage(Stat.SPD)).toBe(-3);
+    }
+  );
+
+  test("does not affect Pokemon with the ability Bulletproof",
```
Comment (DayKev) [inline src/test/moves/syrup_bomb.test.ts:63]:
```suggestion
  it("does not affect Pokemon with the ability Bulletproof",
```

File: src/test/moves/syrup_bomb.test.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    async() => {
+      game.override.enemyAbility(Abilities.BULLETPROOF);
+      await game.startBattle([Species.MAGIKARP]);
+
+      const targetPokemon = game.scene.getEnemyPokemon()!;
+
+      game.move.select(Moves.SYRUP_BOMB);
+      await game.toNextTurn();
+      expect(targetPokemon.getMaxHp()).toBe(targetPokemon.hp);
```
Comment (DayKev) [inline src/test/moves/syrup_bomb.test.ts:72]:
```suggestion
      expect(targetPokemon.isFullHp()).toBe(true);
```

File: src/test/moves/syrup_bomb.test.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      .enemyLevel(100)
+      .moveset([Moves.SYRUP_BOMB, Moves.SPLASH])
+      .enemyMoveset(Moves.SPLASH);
+    vi.spyOn(allMoves[Moves.SYRUP_BOMB], "accuracy", "get").mockReturnValue(100);
+  });
+
+  //Bulbapedia Reference: https://bulbapedia.bulbagarden.net/wiki/syrup_bomb_(move)
+
+  test("decreases the target Pokemon's speed stat once per turn for 3 turns",
+    async() => {
```
Comment (DayKev) [inline src/test/moves/syrup_bomb.test.ts:40]:
```suggestion
    async () => {
```

File: src/test/moves/syrup_bomb.test.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+      game.move.select(Moves.SPLASH);
+      await game.toNextTurn();
+      expect(targetPokemon.getTag(BattlerTagType.SYRUP_BOMB)).toBeUndefined();
+      expect(targetPokemon.getStatStage(Stat.SPD)).toBe(-3);
+    }
+  );
+
+  test("does not affect Pokemon with the ability Bulletproof",
+    async() => {
```
Comment (DayKev) [inline src/test/moves/syrup_bomb.test.ts:64]:
```suggestion
    async () => {
```

File: src/data/battler-tags.ts [hunk 8]
```diff
@@ -2589,6 +2589,45 @@ export class ImprisonTag extends MoveRestrictionBattlerTag {
...
+ * For three turns, starting from the turn of hit, at the end of each turn, the target Pokemon's speed will decrease by 1.
+ * The tag can also expire by taking the target Pokemon off the field.
+ */
+export class SyrupBombTag extends BattlerTag {
+  constructor() {
+    super(BattlerTagType.SYRUP_BOMB, BattlerTagLapseType.TURN_END, 3, Moves.SYRUP_BOMB);
+  }
+
+  /**
+   * Adds the Syrup Bomb battler tag to the target Pokemon
```
Comment (DayKev) [inline src/data/battler-tags.ts:2603]:
```suggestion
   * Adds the Syrup Bomb battler tag to the target Pokemon.
```

File: src/data/battler-tags.ts [hunk 9]
```diff
@@ -2589,6 +2589,45 @@ export class ImprisonTag extends MoveRestrictionBattlerTag {
   }
 }
 
+/**
+ * Battler Tag that applies the effects of Syrup Bomb to the target Pokemon
```
Comment (DayKev) [inline src/data/battler-tags.ts:2593]:
```suggestion
 * Battler Tag that applies the effects of Syrup Bomb to the target Pokemon.
```

File: src/data/battler-tags.ts [hunk 10]
```diff
@@ -2589,6 +2589,45 @@ export class ImprisonTag extends MoveRestrictionBattlerTag {
...
+ * The tag can also expire by taking the target Pokemon off the field.
+ */
+export class SyrupBombTag extends BattlerTag {
+  constructor() {
+    super(BattlerTagType.SYRUP_BOMB, BattlerTagLapseType.TURN_END, 3, Moves.SYRUP_BOMB);
+  }
+
+  /**
+   * Adds the Syrup Bomb battler tag to the target Pokemon
+   * In addition, this also contains a check for a pre-existing Syrup Bomb tag on the target Pokemon to determine whether the tag should be applied or not
```
Comment (DayKev) [inline src/data/battler-tags.ts:2604]: Isn't there an `onOverlap()` method for that?

Comment (frutescens) [reply src/data/battler-tags.ts:2604]: ah you're right. i'll remove the check then.

**Review by Tempo-anon — COMMENTED** [review]
Are there messages for the syrup activating each turn or when the syrup disappears in-game?

**Comment by frutescens** [comment]
No, I don't think so.

**Review by DayKev — APPROVED** [review]

File: src/data/battler-tags.ts [hunk 11]
```diff
@@ -2589,6 +2589,43 @@ export class ImprisonTag extends MoveRestrictionBattlerTag {
...
+   */
+  override lapse(pokemon: Pokemon, _lapseType: BattlerTagLapseType): boolean {
+    if (!pokemon.isActive(true)) {
+      return false;
+    }
+    pokemon.scene.unshiftPhase(new StatStageChangePhase(
+      pokemon.scene, pokemon.getBattlerIndex(), true,
+      [Stat.SPD], -1, true, false, true
+    ));
+    return --this.turnCount > 0;
```
Comment (torranx) [inline src/data/battler-tags.ts:2626]: nit: can we apply KISS here?

File: src/test/moves/syrup_bomb.test.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  beforeEach(() => {
+    game = new GameManager(phaserGame);
+    game.override
+      .starterSpecies(Species.MAGIKARP)
+      .enemySpecies(Species.SNORLAX)
+      .startingLevel(30)
+      .enemyLevel(100)
+      .moveset([Moves.SYRUP_BOMB, Moves.SPLASH])
+      .enemyMoveset(Moves.SPLASH);
+    vi.spyOn(allMoves[Moves.SYRUP_BOMB], "accuracy", "get").mockReturnValue(100);
```
Comment (torranx) [inline src/test/moves/syrup_bomb.test.ts:34]: nit: helper function `forceHit` exists for this

File: src/data/battler-tags.ts [hunk 13]
```diff
@@ -2589,6 +2589,43 @@ export class ImprisonTag extends MoveRestrictionBattlerTag {
...
+ * The tag can also expire by taking the target Pokemon off the field.
+ */
+export class SyrupBombTag extends BattlerTag {
+  constructor() {
+    super(BattlerTagType.SYRUP_BOMB, BattlerTagLapseType.TURN_END, 3, Moves.SYRUP_BOMB);
+  }
+
+  /**
+   * Adds the Syrup Bomb battler tag to the target Pokemon.
+   * In addition, this also contains a check for a pre-existing Syrup Bomb tag on the target Pokemon to determine whether the tag should be applied or not
```
Comment (torranx) [inline src/data/battler-tags.ts:2604]: obsolete comment?

File: src/data/battler-tags.ts (hunk 11 above)
Comment (frutescens) [reply src/data/battler-tags.ts:2626]: What is KISS?

Comment (Tempo-anon) [reply src/data/battler-tags.ts:2626]: Keeping it simple. I admit that `return --this.turnCount > 0;` is not the easiest thing to understand at a glance

File: src/test/moves/syrup_bomb.test.ts (hunk 12 above)
Comment (frutescens) [reply src/test/moves/syrup_bomb.test.ts:34]: The move misses when I use forceHit

File: src/data/battler-tags.ts (hunk 11 above)
Comment (frutescens) [reply src/data/battler-tags.ts:2625]: it's the same format as the parent class' lapse()

File: src/test/moves/syrup_bomb.test.ts (hunk 12 above)
Comment (frutescens) [reply src/test/moves/syrup_bomb.test.ts:35]: Nevermind, figured it out and fixed.

File: src/data/battler-tags.ts (hunk 11 above)
Comment (frutescens) [reply src/data/battler-tags.ts:2625]: image

**Review by DayKev — APPROVED** [review]

**Pull request merged**